### ISSUE TITLE
[Bug]: Silent garbage output on GPUs in Confidential Computing mode: V2 model runner's UVA views of pinned host memory are stale under CC (works with VLLM_USE_V2_MODEL_RUNNER=0)

### ISSUE THREAD

**Issue by yeighta** [vllm-project/vllm#57224]
### Your current environment

<details>
<summary>Environment (TDX guest with an H100 in Confidential Computing mode)</summary>

```text
OS                           : Ubuntu 26.04.1 LTS (x86_64), kernel 7.0.0-31-generic (Intel TDX guest, 16 vCPU / 64 GB)
Python version               : 3.12.14
PyTorch version              : 2.13.0+cu130
CUDA used to build PyTorch   : 13.0
Nvidia driver version        : 595.91.07 (Canonical-signed open kernel modules)
GPU 0                        : NVIDIA H100 NVL (PCIe passthrough via vfio-pci / iommufd)
vLLM Version                 : 0.29.0 (PyPI wheel)
flashinfer                   : 0.6.18

$ nvidia-smi conf-compute -q
CC State                   : ON
Multi-GPU Mode             : None
CPU CC Capabilities        : INTEL TDX
$ nvidia-smi conf-compute -grs
Confidential Compute GPUs Ready state: ready

Host: Ubuntu 26.04.1, kernel 7.0.0-31 (kvm_intel.tdx=1), QEMU 10.2.1, GPU switched to CC mode with
gpu-admin-tools (--set-cc-mode=on). The same guest runs transformers / CuPy / plain torch correctly.
```
</details>

### 🐛 Describe the bug

On a GPU in NVIDIA Confidential Computing (CC) mode inside a TDX guest, vLLM 0.29.0 with the default (V2) GPU model runner **starts fine and serves requests, but every response is garbage**, for example `!!!!!!!!` (token id 0 repeated) or `[][][][]`. There is no error or warning anywhere, and the same model produces correct output with plain `transformers` on the same GPU, so this is easy to misattribute to the GPU or the CC setup. Setting `VLLM_USE_V2_MODEL_RUNNER=0` makes the output correct.

**Root cause (verified):** the V2 model runner feeds `input_ids`, positions, block tables etc. to the GPU through UVA views of pinned host memory (`vllm/v1/worker/gpu/buffer_utils.py` → `UvaBuffer` → `get_accelerator_view_from_cpu_tensor` → `torch.ops._C.get_cuda_view_from_cpu_tensor`). Under GPU CC the device cannot access guest memory directly (all host/device traffic goes through the driver's encrypted bounce buffers), and the UVA view **does not follow later host writes**: the GPU keeps reading the contents the buffer had when it was mapped. The model therefore sees all-zero `input_ids` and confidently predicts token 0 at every position. Prompt logprobs make this visible: with `prompt_logprobs=1` every prompt position predicts `!` with logprob ≈ 0.

Minimal reproduction of the UVA behaviour (no model needed), run inside the CC guest:

```python
import torch, vllm  # importing vllm loads the _C ops
cpu = torch.arange(1, 9, dtype=torch.int32).pin_memory()
gpu_view = torch.ops._C.get_cuda_view_from_cpu_tensor(cpu)
torch.cuda.synchronize()
print("initial          cpu:", cpu.tolist(), " gpu view:", (gpu_view + 0).tolist())
cpu.fill_(99); torch.cuda.synchronize()
print("after host write cpu:", cpu.tolist(), " gpu view:", (gpu_view + 0).tolist())
dst = torch.empty_like(cpu, device="cuda"); dst.copy_(cpu, non_blocking=True); torch.cuda.synchronize()
print("explicit H2D copy       :", dst.tolist())
```

Output on the CC GPU:

```text
initial          cpu: [1, 2, 3, 4, 5, 6, 7, 8]  gpu view: [1, 2, 3, 4, 5, 6, 7, 8]
after host write cpu: [99, 99, 99, 99, 99, 99, 99, 99]  gpu view: [1, 2, 3, 4, 5, 6, 7, 8]   <-- stale
explicit H2D copy       : [99, 99, 99, 99, 99, 99, 99, 99]
```

Ordinary pinned and pageable H2D/D2H copies (blocking and `non_blocking=True`) are all correct on the same GPU; only the zero-copy UVA view is stale.

End-to-end reproduction:

```bash
# default runner (V2): garbage
vllm serve Qwen/Qwen2.5-0.5B-Instruct --enforce-eager --max-model-len 2048 --dtype bfloat16 --attention-backend FLASH_ATTN
curl -s localhost:8000/v1/completions -H 'Content-Type: application/json' \
  -d '{"model":"Qwen/Qwen2.5-0.5B-Instruct","prompt":"The capital of France is","max_tokens":8,"temperature":0}'
# -> "text": "!!!!!!!!"      (also with --attention-backend TRITON_ATTN and with --compilation-config '{"custom_ops":["none"]}')

# V1 runner: correct
VLLM_USE_V2_MODEL_RUNNER=0 vllm serve Qwen/Qwen2.5-0.5B-Instruct --enforce-eager --max-model-len 2048 --dtype bfloat16
# -> "text": " Paris. It was founded in 7"
```

**Suggested fix:** treat UVA as unavailable when the GPU is in Confidential Computing mode, so the V2 runner is not selected (or falls back to explicit copies). CC mode can be detected with NVML (`nvmlSystemGetConfComputeState` / `nvmlSystemGetConfComputeSettings`, `nvidia-smi conf-compute -q`), or `is_uva_available()` could run a tiny write-after-map coherence check like the snippet above at startup and disable UVA when the view is stale. Given that the V1 runner is the only working path today, a loud warning or error would already be a big improvement over silently corrupted output. Related CC-mode limitation (cudaHostRegister) is tracked in #55428.

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

**Comment by devtyagi3909** [comment]
i tracked this down to the zero-copy view not syncing host writes under cc. working on a pr to run a tiny coherence check in `is_uva_available()` and fall back if it fails.

**Comment by yeighta** [comment]
Cross-link: the same `VLLM_USE_V2_MODEL_RUNNER=0` switch that fixes the garbage output here also fixes two other reports of mine on the same hardware, which I had been treating as unrelated:

- **#57562** — `AsyncScheduler` `num_output_placeholders` underflow killing EngineCore. 42 underflow events under a 12-request load with the V2 runner, **0** with V1, same machine, same load.
- **#57680** — decode throughput 3.3x lower on 0.29.0 than 0.26.0. 41/61 tok/s bimodal at c=1 with V2, **151.4 tok/s stable** with V1, against 148.3 for 0.26.0.

Measured on Azure NCC H100 NVL 94GB (SEV-SNP + H100 CC mode), vLLM 0.29.0, `Qwen/Qwen3.6-35B-A3B-FP8`. Details and the full table are in #57562.

If the stale UVA views of pinned host memory described in this issue also cover the sampled-token buffer and the bookkeeping the scheduler reads back, that would account for all three symptoms at once: wrong token ids, tokens appearing for steps nothing was reserved for, and enough extra synchronisation cost to show up as a 3x decode slowdown on the CC link. I have not verified that — it is a hypothesis that fits three measurements.
